src/artifactContentFileSystemProvider.ts

```diff
@@ -0,0 +1,333 @@
+        }
+
+        throw new Error(`There is no contents that can be rendered associated with artifact index ${artifactContent.parsedUriData.runIndex} for run index ${artifactContent.parsedUriData.runIndex}.`);
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public createDirectory(uri: vscode.Uri): void {
+        throw new Error('Not implemented');
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public writeFile(uri: vscode.Uri, content: Uint8Array, options: { create: boolean; overwrite: boolean }): void {
+        throw new Error('Not implemented');
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public delete(uri: vscode.Uri, options: { recursive: boolean }): void {
+        throw new Error('Not implemented');
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public rename(oldUri: vscode.Uri, newUri: vscode.Uri, options: { overwrite: boolean }): void {
+        throw new Error('Not implemented');
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public dispose(): void {
+        vscode.Disposable.from(...this.disposables).dispose();
+        this.disposables = [];
+    }
+
+    /**
+     * Creates an embedded content URI.
+     * @param sarifLog The raw sarif log.
+     * @param logPath The full path to the SARIF log file.
+     * @param artifactUri The file name that VSCode will display in the editor and use for detection of type.
+     * @param runIndex The index of the run in the SARIF file.
+     * @param artifactIndex The artifact index.
+     * @param requiredExtension Indicates how to render binary content.
+     */
+    public static tryCreateUri(sarifLog: sarif.Log, logPath: vscode.Uri, artifactUri: vscode.Uri, runIndex: number, artifactIndex: number, requiredExtension: string | undefined): vscode.Uri | undefined {
+        if (!logPath.isSarifFile()) {
+            throw new Error(`${logPath.toString()} is not a SARIF file`);
+        }
+
+        const run: sarif.Run | undefined = sarifLog.runs[runIndex];
+        if (!run || !run.artifacts) {
+            return undefined;
+        }
+        const artifact: sarif.Artifact | undefined = run.artifacts[artifactIndex];
+        if (!artifact || !artifact.contents) {
+            return undefined;
+        }
+
+        let uriPath: string;
+        if (!requiredExtension) {
+            uriPath = artifactUri.path;
+        } else {
+            const vscodeUri: vscode.Uri = artifactUri.with({ path: artifactUri.path.concat(requiredExtension) });
+            uriPath = path.posix.join(artifactUri.path, vscodeUri.path);
+        }
+
+        const encodedUriData: EncodedUriData = {
+            artifactIndex,
+            runIndex,
+            log: logPath.toString(/*skipEncoding*/ true),
+            artifactUri: artifactUri.toString(/*skpEncoding*/ true)
+        };
+
+        const encodedUriDataAsBase64: string = new Buffer(JSON.stringify(encodedUriData)).toString('base64');
```

Comment:
Could we have the URI as human readable as possible for debugging purposes? Index values (numeric) are URI safe and can be transparent. String values can be URI encoded for maximum readability. You can use the path split and store these values.

Reply:
This pattern (using base64) was used explicitly so I did not have to implement my own parsing.

Writing your own string parsing should at all costs be avoided as it is super easy to miss edge cases and in some cases cause security issues.

Using a well tested encoder\decoder for this is a better choice.

---
In reply to: [411819043](https://github.com/microsoft/sarif-vscode-extension/pull/252#discussion_r411819043) [](ancestors = 411819043)

Reply:
Propose: sarif:/logUri/runIndex/artifactIndex/file.md

Reply:
Noting parsing as a potential security risk (Gabe).

Reply:
We can consider this based on feedback from users since these URIs show up when you hover over the tab. Having the URI most relevant to the artifact is, in my opinion, better

---
In reply to: [412415505](https://github.com/microsoft/sarif-vscode-extension/pull/252#discussion_r412415505) [](ancestors = 412415505)